File: src/sim_workbench/fcb_simulator/src/mmg_model.cpp
// SPDX-License-Identifier: Proprietary
// 4-DOF MMG equations of motion — Yasukawa & Yoshimura 2015 [R7].
#include "fcb_simulator/mmg_model.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <string>

namespace fcb_sim {

namespace {

constexpr double kVelocityFloor = 1e-6;  // m/s, divide-by-zero guard for U
constexpr double kPi = 3.14159265358979323846;

// Compute hull hydrodynamic forces in body frame.
// Returns {X_H, Y_H, N_H} (N, N, N·m).
struct HullForces { double X; double Y; double N; };

HullForces compute_hull_forces(double u, double v, double r, const MmgParams& p) {
  const double U = std::max(std::sqrt(u * u + v * v), kVelocityFloor);
  const double v_p = v / U;
  const double r_p = r * p.L / U;

  const double X_nd = p.X_vv * v_p * v_p
                    + p.X_vr * v_p * r_p
                    + p.X_rr * r_p * r_p
                    + p.X_vvvv * v_p * v_p * v_p * v_p;
  const double Y_nd = p.Y_v * v_p
                    + p.Y_r * r_p
                    + p.Y_vvv * v_p * v_p * v_p
                    + p.Y_vvr * v_p * v_p * r_p
                    + p.Y_vrr * v_p * r_p * r_p
                    + p.Y_rrr * r_p * r_p * r_p;
  const double N_nd = p.N_v * v_p
                    + p.N_r * r_p
                    + p.N_vvv * v_p * v_p * v_p
                    + p.N_vvr * v_p * v_p * r_p
                    + p.N_vrr * v_p * r_p * r_p
                    + p.N_rrr * r_p * r_p * r_p;

  const double q = 0.5 * p.rho * p.L * p.d * U * U;
  HullForces hf;
  hf.X = q * X_nd;
  hf.Y = q * Y_nd;
  hf.N = q * p.L * N_nd;
  return hf;
}

// Propeller thrust X_P (N).
double compute_thrust(double u, double n_rps, const MmgParams& p) {
  if (std::abs(n_rps) < 1e-3) {
    return 0.0;
  }
  const double J = u * (1.0 - p.w_P) / (n_rps * p.D_P);
  const double K_T = p.k_2 * J * J + p.k_1 * J + p.k_0;
  return (1.0 - p.t_P) * p.rho * n_rps * n_rps
         * std::pow(p.D_P, 4) * K_T;
}

struct RudderForces { double X; double Y; double N; };

RudderForces compute_rudder_forces(double u, double v, double r,
                                   double delta, double n_rps,
                                   const MmgParams& p) {
  const double U = std::max(std::sqrt(u * u + v * v), kVelocityFloor);
  const double beta = std::atan2(-v, std::max(u, kVelocityFloor));
  const double r_p = r * p.L / U;

  // Effective inflow at rudder
  const double v_R = U * p.gamma_R * (beta - p.l_R_prime * r_p);

  // C_Th — propeller thrust loading coefficient
  double C_Th = 0.0;
  if (std::abs(n_rps) > 1e-3 && std::abs(u) > kVelocityFloor) {
    const double thrust = compute_thrust(u, n_rps, p);
    const double denom = 0.5 * p.rho * (kPi * 0.25 * p.D_P * p.D_P)
                         * u * u;
    C_Th = std::min(thrust / std::max(denom, 1e-6), 20.0);  // cap prevents blow-up below ~0.03 m/s
  }

  const double eta = p.D_P / p.d;  // propeller-to-rudder height ratio (approx)
  const double bracket = eta
      * std::pow(1.0 + p.kappa * (std::sqrt(std::max(1.0 + C_Th, 0.0)) - 1.0), 2)
      + (1.0 - eta);
  const double u_R = p.epsilon * u * std::sqrt(std::max(bracket, 0.0));

  const double U_R = std::sqrt(u_R * u_R + v_R * v_R);
  const double alpha_R = delta - std::atan2(v_R, std::max(u_R, kVelocityFloor));
  const double F_N = 0.5 * p.rho * p.A_R * p.f_alpha * U_R * U_R
                     * std::sin(alpha_R);

  RudderForces rf;
  rf.X = -(1.0 - p.t_R) * F_N * std::sin(delta);
  rf.Y = -(1.0 + p.a_H) * F_N * std::cos(delta);
  rf.N = -(p.x_R_prime + p.a_H * p.x_H_prime) * p.L
         * F_N * std::cos(delta);
  return rf;
}

}  // namespace

FcbState compute_derivatives(const FcbState& s,
                             double delta_rad,
                             double n_rps,
                             const MmgParams& p) {
  // Mass + added mass (dimensional)
  const double mass = p.displacement_t * 1000.0;
  const double half_rho_L2_d = 0.5 * p.rho * p.L * p.L * p.d;
  const double m_x = p.m_x_prime * half_rho_L2_d;
  const double m_y = p.m_y_prime * half_rho_L2_d;
  const double I_zz = mass * std::pow(0.25 * p.L, 2);  // approx radius of gyration L/4
  const double J_zz = p.J_zz_prime * 0.5 * p.rho
                      * std::pow(p.L, 4) * p.d;

  const HullForces hf = compute_hull_forces(s.u, s.v, s.r, p);
  const RudderForces rf = compute_rudder_forces(s.u, s.v, s.r,
                                                delta_rad, n_rps, p);
  const double X_P = compute_thrust(s.u, n_rps, p);

  const double X = hf.X + rf.X + X_P;
  const double Y = hf.Y + rf.Y;
  const double N = hf.N + rf.N;

  // Solve the coupled (v_dot, r_dot) 2x2 system:
  //   (m + m_y) v_dot + m * x_G * r_dot = Y - (m + m_x) u r
  //   m * x_G * v_dot + (I_zz + J_zz) r_dot = N - m * x_G * u r
  const double a11 = mass + m_y;
  const double a12 = mass * p.x_G;
  const double a21 = mass * p.x_G;
  const double a22 = I_zz + J_zz;
  const double rhs1 = Y - (mass + m_x) * s.u * s.r;
  const double rhs2 = N - mass * p.x_G * s.u * s.r;
  const double det = a11 * a22 - a12 * a21;
  if (std::abs(det) < 1e-6) {
    throw std::runtime_error("MMG inertia matrix near-singular (det=" + std::to_string(det) +
                             "); check x_G in YAML");
  }
  const double v_dot = (a22 * rhs1 - a12 * rhs2) / det;
  const double r_dot = (-a21 * rhs1 + a11 * rhs2) / det;

  // Surge:
  //   (m + m_x) u_dot = X + (m + m_y) v r + x_G m r^2
  const double u_dot = (X + (mass + m_y) * s.v * s.r
                        + p.x_G * mass * s.r * s.r) / (mass + m_x);

  // Roll (1-DOF linear pendulum, weakly coupled):
  //   phi_ddot = -omega_n^2 * phi - 2*zeta*omega_n * phi_dot  (linearised 1-DOF; sway coupling deferred to HAZID)
  // G_M exposed in YAML for HAZID calibration; omega_n derived from T_phi only until roll model matures
  const double omega_n = 2.0 * kPi / p.T_phi;
  constexpr double kZeta = 0.05;
  const double phi_ddot = -omega_n * omega_n * s.phi
                          - 2.0 * kZeta * omega_n * s.phi_dot;

  // Kinematics (NED, flat earth):
  //   x_dot = u cos(psi) - v sin(psi)
  //   y_dot = u sin(psi) + v cos(psi)
  //   psi_dot = r
  FcbState d;
  d.x = s.u * std::cos(s.psi) - s.v * std::sin(s.psi);
  d.y = s.u * std::sin(s.psi) + s.v * std::cos(s.psi);
  d.psi = s.r;
  d.u = u_dot;
  d.v = v_dot;
  d.r = r_dot;
  d.phi = s.phi_dot;
  d.phi_dot = phi_ddot;
  return d;
}

}  // namespace fcb_sim

File: src/sim_workbench/sil_nodes/ship_dynamics/ship_dynamics/mmg_model.py
"""Yasukawa & Yoshimura 2015 4-DOF MMG 模型 — RK4 积分器。

参考文献:
  Yasukawa, H. & Yoshimura, Y., "Introduction of MMG standard method
  for ship maneuvering predictions", J. Mar. Sci. Tech., 2015 [R7]

运动方程 (船体固定坐标系, CG 原点, x_G=0):
  m11*u_dot - m22*v*r = X_total
  m22*v_dot + m11*u*r = Y_total
  m33*r_dot           = N_total
  m44*p_dot           = K_total

其中:
  m11 = mass*(1+m_x'), m22 = mass*(1+m_y'), m33 = I_zz, m44 = I_xx
"""

import math
from dataclasses import dataclass
from .mmg_coefficients import MMGCoefficients


@dataclass
class ShipState:
    """8 自由度状态向量: [x, y, psi, phi, u, v, r, p]"""

    x: float = 0.0
    y: float = 0.0
    psi: float = 0.0
    phi: float = 0.0
    u: float = 0.0
    v: float = 0.0
    r: float = 0.0
    p: float = 0.0

    def __add__(self, other: "ShipState") -> "ShipState":
        return ShipState(
            x=self.x + other.x,
            y=self.y + other.y,
            psi=self.psi + other.psi,
            phi=self.phi + other.phi,
            u=self.u + other.u,
            v=self.v + other.v,
            r=self.r + other.r,
            p=self.p + other.p,
        )

    def __mul__(self, scalar: float) -> "ShipState":
        return ShipState(
            x=self.x * scalar,
            y=self.y * scalar,
            psi=self.psi * scalar,
            phi=self.phi * scalar,
            u=self.u * scalar,
            v=self.v * scalar,
            r=self.r * scalar,
            p=self.p * scalar,
        )

    __rmul__ = __mul__


class MMGModel:
    """Yasukawa 2015 4-DOF MMG 动力学模型 + RK4 积分器。

    状态:  [x, y, psi, phi, u, v, r, p]
    控制:  [delta_cmd (rad), n_rps_cmd (rev/s)]
    环境:  [wind_speed, wind_dir, current_speed, current_dir]
    """

    def __init__(self, coeffs: MMGCoefficients):
        self.c = coeffs
        self._precompute_mass_matrix()

    def _precompute_mass_matrix(self):
        c = self.c
        self.m11 = c.mass * (1.0 + c.m_x_prime)
        self.m22 = c.mass * (1.0 + c.m_y_prime)
        self.m33 = c.I_zz
        self.m44 = c.I_xx

    # ─── 主接口 ───────────────────────────────────────────────

    def rk4_step(
        self,
        state: ShipState,
        delta_cmd: float,
        n_rps_cmd: float,
        wind_speed: float = 0.0,
        wind_dir_rad: float = 0.0,
        current_speed: float = 0.0,
        current_dir_rad: float = 0.0,
    ) -> ShipState:
        """4 阶 Runge-Kutta 推进一个时间步 (dt)。"""
        dt = self.c.dt

        k1 = self.compute_derivatives(
            state, delta_cmd, n_rps_cmd,
            wind_speed, wind_dir_rad, current_speed, current_dir_rad,
        )
        k2 = self.compute_derivatives(
            state + k1 * (0.5 * dt),
            delta_cmd, n_rps_cmd,
            wind_speed, wind_dir_rad, current_speed, current_dir_rad,
        )
        k3 = self.compute_derivatives(
            state + k2 * (0.5 * dt),
            delta_cmd, n_rps_cmd,
            wind_speed, wind_dir_rad, current_speed, current_dir_rad,
        )
        k4 = self.compute_derivatives(
            state + k3 * dt,
            delta_cmd, n_rps_cmd,
            wind_speed, wind_dir_rad, current_speed, current_dir_rad,
        )

        return state + (k1 + k2 * 2.0 + k3 * 2.0 + k4) * (dt / 6.0)

    def compute_derivatives(
        self,
        state: ShipState,
        delta_cmd: float,
        n_rps_cmd: float,
        wind_speed: float = 0.0,
        wind_dir_rad: float = 0.0,
        current_speed: float = 0.0,
        current_dir_rad: float = 0.0,
    ) -> ShipState:
        """计算完整的状态导数 d(state)/dt。"""
        c = self.c

        # ── (a) 运动学 ──
        x_dot = state.u * math.cos(state.psi) - state.v * math.sin(state.psi)
        y_dot = state.u * math.sin(state.psi) + state.v * math.cos(state.psi)
        psi_dot = state.r
        phi_dot = state.p

        # ── (b) 无因次化 ──
        U = math.sqrt(state.u ** 2 + state.v ** 2)
        if U < 1e-6:
            U = 1e-6
        v_prime = state.v / U
        r_prime = state.r * c.L / U

        # ── (c) 裸船体力 ──
        X_hull_nd = (
            c.X_vv * v_prime ** 2
            + c.X_vr * v_prime * r_prime
            + c.X_rr * r_prime ** 2
            + c.X_vvvv * v_prime ** 4
        )
        Y_hull_nd = (
            c.Y_v * v_prime
            + c.Y_r * r_prime
            + c.Y_vvv * v_prime ** 3
            + c.Y_vvr * v_prime ** 2 * r_prime
            + c.Y_vrr * v_prime * r_prime ** 2
            + c.Y_rrr * r_prime ** 3
        )
        N_hull_nd = (
            c.N_v * v_prime
            + c.N_r * r_prime
            + c.N_vvv * v_prime ** 3
            + c.N_vvr * v_prime ** 2 * r_prime
            + c.N_vrr * v_prime * r_prime ** 2
            + c.N_rrr * r_prime ** 3
        )
        U2 = U * U
        X_hull = X_hull_nd * c.scale_force * U2
        Y_hull = Y_hull_nd * c.scale_force * U2
        N_hull = N_hull_nd * c.scale_moment * U2

        # ── (d) 螺旋桨推力 ──
        X_prop = 0.0
        if n_rps_cmd >= 0.01:
            J = state.u * (1.0 - c.w_P) / (n_rps_cmd * c.D_P)
            K_T = c.k_2 * J * J + c.k_1 * J + c.k_0
            X_prop = (1.0 - c.t_P) * c.rho * n_rps_cmd ** 2 * c.D_P ** 4 * K_T

        # ── (e) 舵力 ──
        u_R = c.epsilon * state.u * (1.0 - c.w_P)
        v_R = c.gamma_R * (state.v + c.l_R_prime * c.L * state.r)
        U_R = math.sqrt(u_R ** 2 + v_R ** 2)
        if U_R < 1e-6:
            U_R = 1e-6
        alpha_R = delta_cmd - math.atan2(v_R, u_R)
        F_N = 0.5 * c.rho * c.A_R * c.f_alpha * U_R * U_R * math.sin(alpha_R)
        X_R = -(1.0 - c.t_R) * F_N * math.sin(delta_cmd)
        Y_R = -(1.0 + c.a_H) * F_N * math.cos(delta_cmd)
        N_R = -((c.x_R_prime + c.a_H * c.x_H_prime) * c.L) * F_N * math.cos(delta_cmd)

        # ── (f) 环境力 (二次阻力简化) ──
        X_env = 0.0
        Y_env = 0.0
        rho_air = 1.225
        wind_area = c.B * (c.L * 0.3)  # 侧投影面积近似
        wind_drag_coeff = 0.8

        if wind_speed > 0.01:
            F_wind = 0.5 * rho_air * wind_area * wind_drag_coeff * wind_speed ** 2
            X_env += F_wind * math.cos(wind_dir_rad - state.psi)
            Y_env += F_wind * math.sin(wind_dir_rad - state.psi)

        if current_speed > 0.01:
            F_current = 0.5 * c.rho * c.L * c.d * 0.05 * current_speed ** 2
            X_env += F_current * math.cos(current_dir_rad - state.psi)
            Y_env += F_current * math.sin(current_dir_rad - state.psi)

        # ── (g & h) 质量矩阵求解 ──
        X_total = X_hull + X_prop + X_R + X_env
        Y_total = Y_hull + Y_R + Y_env
        N_total = N_hull + N_R

        u_dot = (X_total + self.m22 * state.v * state.r) / self.m11
        v_dot = (Y_total - self.m11 * state.u * state.r) / self.m22
        r_dot = N_total / self.m33

        # ── (i) 横摇 (线性 1-DOF) ──
        K_phi = c.mass * 9.81 * c.G_M
        K_p = K_phi * c.T_phi / (2.0 * math.pi) * 0.1
        p_dot = (-K_phi * state.phi - K_p * state.p) / self.m44

        return ShipState(
            x=x_dot, y=y_dot, psi=psi_dot, phi=phi_dot,
            u=u_dot, v=v_dot, r=r_dot, p=p_dot,
        )
